tests/test_dynamics.py

```diff
@@ -194,6 +291,241 @@ def test_torque_derivative_driven(with_contact, with_external_force, cx):
+            [
+                0.6118529,
+                0.785176,
+                0.6075449,
+                0.8083973,
+                -0.3024137,
+                -10.3850379,
+                1.6044517,
+                35.8023864,
+                0.3886773,
+                0.5426961,
+                0.7722448,
+                0.7290072,
+            ],
+        )
+
+
+@pytest.mark.parametrize(
+    "dynamics",
+    [DynamicsFcn.TORQUE_ACTIVATIONS_DRIVEN, DynamicsFcn.MUSCLE_DRIVEN],
+)
+def test_implicit_soft_contacts_errors(dynamics):
+    # Prepare the program
+    nlp = NonLinearProgram()
+    nlp.model = biorbd.Model(
+        TestUtils.bioptim_folder() + "/examples/getting_started/models/2segments_4dof_2contacts.bioMod"
+    )
+    nlp.ns = 5
+    nlp.cx = MX
+
+    nlp.u_bounds = np.zeros((nlp.model.nbQ() * 4, 1))
+    ocp = OptimalControlProgram(nlp)
+    nlp.control_type = ControlType.CONSTANT
+    NonLinearProgram.add(
+        ocp,
+        "dynamics_type",
+        Dynamics(dynamics, implicit_soft_contacts=True),
+        False,
+    )
+
+    # Prepare the dynamics
+    if dynamics == DynamicsFcn.TORQUE_DERIVATIVE_DRIVEN:
+        with pytest.raises(
+            TypeError,
+            match=re.escape(
+                "torque_activations_driven() got an unexpected keyword argument " "'implicit_soft_contacts'"
+            ),
+        ):
+            ConfigureProblem.initialize(ocp, nlp)
+    elif dynamics == DynamicsFcn.MUSCLE_DRIVEN:
+        with pytest.raises(
+            TypeError,
+            match=re.escape("muscle_driven() got an unexpected keyword argument " "'implicit_soft_contacts'"),
+        ):
+            ConfigureProblem.initialize(ocp, nlp)
+
+
+@pytest.mark.parametrize(
+    "dynamics",
+    [DynamicsFcn.TORQUE_ACTIVATIONS_DRIVEN, DynamicsFcn.MUSCLE_DRIVEN],
+)
+def test_implicit_dynamics_errors(dynamics):
+    # Prepare the program
+    nlp = NonLinearProgram()
+    nlp.model = biorbd.Model(
+        TestUtils.bioptim_folder() + "/examples/getting_started/models/2segments_4dof_2contacts.bioMod"
+    )
+    nlp.ns = 5
+    nlp.cx = MX
+
+    nlp.u_bounds = np.zeros((nlp.model.nbQ() * 4, 1))
+    ocp = OptimalControlProgram(nlp)
+    nlp.control_type = ControlType.CONSTANT
+    NonLinearProgram.add(
+        ocp,
+        "dynamics_type",
+        Dynamics(dynamics, implicit_dynamics=True),
+        False,
+    )
+
+    # Prepare the dynamics
+    if dynamics == DynamicsFcn.TORQUE_DERIVATIVE_DRIVEN:
+        with pytest.raises(
+            TypeError,
+            match=re.escape("torque_activations_driven() got an unexpected keyword argument " "'implicit_dynamics'"),
+        ):
+            ConfigureProblem.initialize(ocp, nlp)
+    elif dynamics == DynamicsFcn.MUSCLE_DRIVEN:
+        with pytest.raises(
+            TypeError,
+            match=re.escape("muscle_driven() got an unexpected keyword argument " "'implicit_dynamics'"),
+        ):
+            ConfigureProblem.initialize(ocp, nlp)
+
+
```

Comment:
Again, this should not raise, the signature of the `Dynamics(...)` should be the exact same for all functions

Reply:
removed if
the error is located higher in a function, we get an error because theses dynamics cannot receive implicit_soft_contacts or implicit_dynamics as an argument.